File: src/core/tiers/l1-cache.ts
/**
 * L1 Transient Cache
 * Fast vector-based memory with TTL eviction
 */

import type { MemoryUnit, Tier } from '../memory/types';
import { now } from '../../lib/utils/time';
import { cosineSimilarity } from '../memory/decoder';

/**
 * L1 Cache configuration
 */
export interface L1CacheConfig {
  /** Maximum entries in cache */
  maxSize: number;
  /** Default TTL in milliseconds */
  defaultTtlMs: number;
  /** Eviction check interval */
  evictionIntervalMs: number;
}

/**
 * Cached entry with TTL
 */
interface CachedEntry {
  unit: MemoryUnit;
  expiresAt: number;
  insertedAt: number;
}

export const DEFAULT_L1_CONFIG: L1CacheConfig = {
  maxSize: 1000,
  defaultTtlMs: 3600000, // 1 hour
  evictionIntervalMs: 60000 // 1 minute
};

/**
 * L1 Transient Cache implementation
 */
export class L1Cache {
  private cache: Map<string, CachedEntry> = new Map();
  private config: L1CacheConfig;
  private evictionTimer?: ReturnType<typeof setInterval>;

  constructor(config: L1CacheConfig = DEFAULT_L1_CONFIG) {
    this.config = config;
  }

  /**
   * Start automatic eviction timer
   */
  startEviction(): void {
    if (this.evictionTimer) return;
    this.evictionTimer = setInterval(
      () => this.evictExpired(),
      this.config.evictionIntervalMs
    );
  }

  /**
   * Stop automatic eviction timer
   */
  stopEviction(): void {
    if (this.evictionTimer) {
      clearInterval(this.evictionTimer);
      this.evictionTimer = undefined;
    }
  }

  /**
   * Insert a memory unit into cache
   */
  insert(unit: MemoryUnit, ttlMs?: number): void {
    const ttl = ttlMs ?? this.config.defaultTtlMs;
    const timestamp = now();

    // Evict oldest if at capacity
    if (this.cache.size >= this.config.maxSize) {
      this.evictOldest();
    }

    this.cache.set(unit.uor_id, {
      unit,
      insertedAt: timestamp,
      expiresAt: timestamp + ttl
    });
  }

  /**
   * Retrieve a unit by UOR ID
   */
  get(uorId: string): MemoryUnit | null {
    const entry = this.cache.get(uorId);
    if (!entry) return null;

    // Check expiration
    if (entry.expiresAt < now()) {
      this.cache.delete(uorId);
      return null;
    }

    return entry.unit;
  }

  /**
   * Find top-K similar units by embedding
   */
  topK(queryEmbedding: Float32Array, k: number): MemoryUnit[] {
    const currentTime = now();
    const scored: Array<{ unit: MemoryUnit; score: number }> = [];

    for (const entry of this.cache.values()) {
      if (entry.expiresAt < currentTime) continue;

      const score = cosineSimilarity(queryEmbedding, entry.unit.embedding);
      scored.push({ unit: entry.unit, score });
    }

    return scored
      .sort((a, b) => b.score - a.score)
      .slice(0, k)
      .map(s => s.unit);
  }

  /**
   * Remove a specific entry
   */
  remove(uorId: string): boolean {
    return this.cache.delete(uorId);
  }

  /**
   * Evict all expired entries
   */
  evictExpired(): number {
    const currentTime = now();
    let evicted = 0;

    for (const [id, entry] of this.cache.entries()) {
      if (entry.expiresAt < currentTime) {
        this.cache.delete(id);
        evicted++;
      }
    }

    return evicted;
  }

  /**
   * Evict oldest entry (LRU-style)
   */
  private evictOldest(): void {
    let oldest: string | null = null;
    let oldestTime = Infinity;

    for (const [id, entry] of this.cache.entries()) {
      if (entry.insertedAt < oldestTime) {
        oldest = id;
        oldestTime = entry.insertedAt;
      }
    }

    if (oldest) {
      this.cache.delete(oldest);
    }
  }

  /**
   * Clear all entries
   */
  clear(): void {
    this.cache.clear();
  }

  /**
   * Get current cache size
   */
  size(): number {
    return this.cache.size;
  }

  /**
   * Check if UOR ID exists in cache
   */
  has(uorId: string): boolean {
    const entry = this.cache.get(uorId);
    if (!entry) return false;
    if (entry.expiresAt < now()) {
      this.cache.delete(uorId);
      return false;
    }
    return true;
  }
}

/**
 * Create a new L1 cache instance
 */
export function createL1Cache(config?: L1CacheConfig): L1Cache {
  return new L1Cache(config);
}

File: crates/evolve-core/src/tiers/l1_cache.rs
use crate::memory::types::{MemoryUnit, UorId};
use std::collections::HashMap;

/// L1 Transient Cache -- ephemeral memory with TTL eviction.
pub struct L1Cache {
    entries: HashMap<UorId, CacheEntry>,
    ttl_ms: i64,
    max_size: usize,
}

struct CacheEntry {
    unit: MemoryUnit,
    inserted_at: i64,
}

impl L1Cache {
    /// Create a cache with the given TTL (milliseconds) and maximum capacity.
    pub fn new(ttl_ms: i64, max_size: usize) -> Self {
        Self {
            entries: HashMap::new(),
            ttl_ms,
            max_size,
        }
    }

    /// Insert or update a memory unit at the given timestamp.
    pub fn insert(&mut self, unit: MemoryUnit, now: i64) {
        self.evict_expired(now);
        if self.entries.len() >= self.max_size {
            self.evict_oldest();
        }
        let id = unit.uor_id;
        self.entries.insert(id, CacheEntry { unit, inserted_at: now });
    }

    /// Retrieve a memory unit by ID, returning `None` if expired.
    pub fn get(&self, id: &UorId, now: i64) -> Option<&MemoryUnit> {
        let entry = self.entries.get(id)?;
        if now - entry.inserted_at > self.ttl_ms {
            return None;
        }
        Some(&entry.unit)
    }

    /// Remove all entries whose TTL has elapsed.
    pub fn evict_expired(&mut self, now: i64) {
        self.entries.retain(|_, e| now - e.inserted_at <= self.ttl_ms);
    }

    /// Number of entries (including potentially expired).
    pub fn len(&self) -> usize {
        self.entries.len()
    }

    /// Whether the cache is empty.
    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    /// Iterate over all non-expired units.
    pub fn iter_units(&self, now: i64) -> impl Iterator<Item = &MemoryUnit> {
        let ttl = self.ttl_ms;
        self.entries
            .values()
            .filter(move |e| now - e.inserted_at <= ttl)
            .map(|e| &e.unit)
    }

    fn evict_oldest(&mut self) {
        if let Some((&oldest_id, _)) = self
            .entries
            .iter()
            .min_by_key(|(_, e)| e.inserted_at)
        {
            self.entries.remove(&oldest_id);
        }
    }
}
